File: android/src/main/java/com/expoalarmmodule/Alarm.java
package com.expoalarmmodule;

import android.os.Build;

import androidx.annotation.RequiresApi;

import com.google.gson.Gson;

import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.List;

import static com.expoalarmmodule.GsonUtil.createSerialize;



public class Alarm implements Cloneable {

    String uid;

    ZonedDateTime date;
    String dateString;

    ArrayList<Integer> days;
    int hour;
    int minutes;

    String title;
    String description;
    boolean repeating;
    boolean active;
    
    /**
     * Show dismiss button in notification. It is false by default.
     */
    boolean showDismiss;

    /**
     * Show snooze button. It is false by default.
     */
    boolean showSnooze;

    /**
     * Snooze interval in minutes
     * Default is 5 minutes
     */
    int snoozeInterval;

    /**
     * Custom texts for buttons of notifications
     */
    String dismissText;
    String snoozeText;

    Alarm(String uid, ArrayList<Integer> days, ZonedDateTime date, int hour, int minutes, boolean showDismiss, boolean showSnooze, int snoozeInterval, String title, String description, boolean repeating, boolean active, String dismissText, String snoozeText) {
        this.uid = uid;
        this.days = days;
        this.hour = hour;
        this.minutes = minutes;
        this.showDismiss = showDismiss;
        this.showSnooze = showSnooze;
        this.snoozeInterval = snoozeInterval;
        this.title = title;
        this.description = description;
        this.repeating = repeating;
        this.active = active;
        this.dismissText = dismissText;
        this.snoozeText = snoozeText;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
          this.date = date;
        }
    }

    List<Date> getDates() {
        List<Date> dates = new ArrayList<>();

        if(date != null) {
          Calendar triggerDate = null;
          if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            triggerDate = GregorianCalendar.from(date);
            dates.add(triggerDate.getTime());
          }
        } else if (days != null) {
          for (int i = 0; i < days.size(); i++) {
            Calendar date = Helper.getDate(days.get(i), hour, minutes);
            dates.add(date.getTime());
          }
        }
        return dates;
    }

    AlarmDates getAlarmDates() {
        return new AlarmDates(uid, getDates());
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    static Alarm fromJson(String json) {
      Alarm alarmTemp = createSerialize().fromJson(json, Alarm.class);
      alarmTemp.date = ZonedDateTime.parse(alarmTemp.dateString);
      return alarmTemp;
    }

    static String toJson(Alarm alarm) {
        alarm.dateString = alarm.date.toString();
        return createSerialize().toJson(alarm);
    }

    public Alarm clone () throws CloneNotSupportedException {
        return (Alarm) super.clone();
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) return true;
        if (!(o instanceof Alarm)) return false;
        Alarm alarm = (Alarm)o;
        return (
                this.hour == alarm.hour &&
                this.minutes == alarm.minutes &&
                this.showDismiss == alarm.showDismiss &&
                this.showSnooze == alarm.showSnooze &&
                this.snoozeInterval == alarm.snoozeInterval &&
                this.dismissText == alarm.dismissText &&
                this.snoozeText == alarm.snoozeText &&
                this.uid.equals(alarm.uid) &&
                this.days.equals(alarm.days) &&
                this.title.equals(alarm.title) &&
                this.description.equals(alarm.description)
        );
    }
}

File: android/src/main/java/expo/modules.alarmemodule/Alarm.kt
package com.alarm

import com.google.gson.Gson

class Alarm internal constructor(var uid: String, days: ArrayList<Integer?>, hour: Int, minutes: Int, snoozeInterval: Int, title: String, description: String, repeating: Boolean, active: Boolean) : Cloneable {
    var days: ArrayList<Integer>
    var hour: Int
    var minutes: Int
    var snoozeInterval: Int
    var title: String
    var description: String
    var repeating: Boolean
    var active: Boolean

    init {
        this.days = days
        this.hour = hour
        this.minutes = minutes
        this.snoozeInterval = snoozeInterval
        this.title = title
        this.description = description
        this.repeating = repeating
        this.active = active
    }

    val dates: Array<Any?>
        get() {
            val dates: Array<Date?> = arrayOfNulls<Date>(days.size())
            for (i in dates.indices) {
                val date: Calendar = Helper.getDate(days.get(i), hour, minutes)
                dates[i] = date.getTime()
            }
            return dates
        }
    val alarmDates: AlarmDates
        get() = AlarmDates(uid, dates)

    @Throws(CloneNotSupportedException::class)
    fun clone(): Alarm {
        return super.clone() as Alarm
    }

    @Override
    override fun equals(o: Object): Boolean {
        if (o === this) return true
        if (o !is Alarm) return false
        val alarm = o as Alarm
        return hour == alarm.hour && minutes == alarm.minutes && snoozeInterval == alarm.snoozeInterval &&
                uid.equals(alarm.uid) &&
                days.equals(alarm.days) &&
                title.equals(alarm.title) &&
                description.equals(alarm.description)
    }

    companion object {
        fun fromJson(json: String?): Alarm {
            return Gson().fromJson(json, Alarm::class.java)
        }

        fun toJson(alarm: Alarm?): String {
            return Gson().toJson(alarm)
        }
    }
}